tpm/tpm.go

```diff
@@ -713,6 +714,104 @@ func OwnerReadSRK(rw io.ReadWriter, ownerAuth digest) ([]byte, error) {
 	return tpmutil.Pack(pk)
 }
 
+// ReadEKCert reads the EKCert from the NVRAM
+func ReadEKCert(rw io.ReadWriter, ownAuth digest) ([]byte, error) {
+	var (
+		data     []byte
+		ekbuf    []byte
+		buff     io.Reader
+		offset   uint32
+		bufSize  uint32
+		tag      uint16
+		certType uint8
+		certSize uint16
+	)
+	const certIndex = 0x1000f000
+	offset = 0
+	data, err := NVReadValue(rw, certIndex, offset, 5, ownAuth)
+	if err != nil {
+		return nil, err
+	}
+	offset = offset + 5
+	buff = bytes.NewReader(data)
+
+	if err := binary.Read(buff, binary.BigEndian, &tag); err != nil {
+		return nil, fmt.Errorf("binary.Read failed: %v", err)
+	}
+	if err := binary.Read(buff, binary.BigEndian, &certType); err != nil {
+		return nil, fmt.Errorf("binary.Read failed: %v", err)
+	}
+	if err := binary.Read(buff, binary.BigEndian, &certSize); err != nil {
+		return nil, fmt.Errorf("binary.Read failed: %v", err)
+	}
+
+	if tag != 0x1001 { // TCG_TAG_PCCLIENT_STORED_CERT
+		return nil, fmt.Errorf("invalid certificate")
+	}
+
+	switch certType {
+	case 0: // TCG_FULL_CERT
+		data, err := NVReadValue(rw, certIndex, offset, 2, ownAuth)
+		if err != nil {
+			return nil, err
+		}
+		bufSize = uint32(certSize) + offset
+		offset = offset + 2
+		buff = bytes.NewReader(data)
+
+		if err := binary.Read(buff, binary.BigEndian, &tag); err != nil {
+			return nil, fmt.Errorf("binary.Read failed: %v", err)
+		}
+
+		if tag != 0x1002 { // TCG_TAG_PCCLIENT_FULL_CERT
+			return nil, fmt.Errorf("certificate type and tag do not match")
+		}
+
+	case 1: // TCG_PARTIAL_SMALL_CERT
+		return nil, fmt.Errorf("certType is not TCG_FULL_CERT: currently do not support partial certs")
+	}
+
+	for offset < bufSize {
+		length := bufSize - offset
+		if length > 128 { // read the cert in 128byte chunks
```

Comment:
Please explain why you're reading 128 bytes at a time and not some other amount

Reply:
I'm actually not 100% certain it should be 128. I took that value from go-tspi. TPMs are limited by how much they can read at a time but I can't find any documentation about what the limit is. I tested it an Infineon TPM on linux and was able to read 1024bytes at a time.